src/eos-updater-poll.c

```diff
@@ -203,6 +203,192 @@ get_fetchers (SourcesConfig *config,
 }
 
 static void
+object_unref0 (gpointer obj)
+{
+  if (obj == NULL)
+    return;
+  g_object_unref (obj);
+}
+
+static GPtrArray *
+get_finders (SourcesConfig          *config,
+             GMainContext           *context,
+             OstreeRepoFinderAvahi **out_finder_avahi)
+{
+  g_autoptr(OstreeRepoFinderAvahi) finder_avahi = NULL;
+  g_autoptr(GPtrArray) finders = g_ptr_array_new_full (config->download_order->len,
+                                                       object_unref0);
+  g_autoptr(GError) local_error = NULL;
+  gsize i;
+
+  g_assert (config->download_order->len > 0);
+
+  for (i = 0; i < config->download_order->len; i++)
+    {
+      switch (g_array_index (config->download_order, EosUpdaterDownloadSource, i))
+        {
+        case EOS_UPDATER_DOWNLOAD_MAIN:
+          g_ptr_array_add (finders, ostree_repo_finder_config_new ());
+          break;
+
+        case EOS_UPDATER_DOWNLOAD_LAN:
+          finder_avahi = ostree_repo_finder_avahi_new (context);
+          g_ptr_array_add (finders, g_object_ref (finder_avahi));
+          break;
+
+        case EOS_UPDATER_DOWNLOAD_VOLUME:
+          /* TODO: How to make this one testable? */
+          g_ptr_array_add (finders, ostree_repo_finder_mount_new (NULL));
+          break;
+
+        default:
+          g_assert_not_reached ();
+        }
+    }
+
+  g_ptr_array_add (finders, NULL);  /* NULL terminator */
+
+  /* TODO: Stop this at some point; think of a better way to store it and
+   * control its lifecycle. */
+  if (finder_avahi != NULL)
+    ostree_repo_finder_avahi_start (OSTREE_REPO_FINDER_AVAHI (finder_avahi),
+                                    &local_error);
+
+  if (local_error != NULL)
+    {
+      g_warning ("Avahi finder failed; removing it: %s", local_error->message);
+      g_ptr_array_remove (finders, finder_avahi);
```

Comment:
Maybe instead of adding `finder_avahi` to the array above and then removing it here, it might be better to capture `finder_avahi` in the variable above and then add it to the array if `ostree_repo_finder_avahi_start` succeeded.

Reply:
I can’t see any advantage in doing that; both approaches are ugly, as far as I see. The whole code needs refactoring, but I didn’t want to do that until the old code paths have been dropped. I’ll add a `FIXME` about that.

Reply:
`FIXME` fixup pushed.